allennlp/semparse/contexts/table_question_context.py

```diff
@@ -0,0 +1,130 @@
+                 column_idx_to_name : Dict[int,str]) -> None:
+        self.cell_values = cell_values
+        self.column_types = { column_idx_to_name[column_idx] : max(column_type_statistics[column_idx]) 
+                              for column_idx in column_idx_to_name }
+            
+       
+
+    @classmethod
+    def parse_file(cls, filename: str, question: List[Token], max_tokens_for_num_cell : int) -> 'TableQuestionKnowledgeGraph':
+        f = open(filename, 'r')
+        reader = csv.reader(f,  delimiter='\t', quoting=csv.QUOTE_NONE)
+        # obtain column information
+        lines = [line for line in reader]
+        column_idx_to_name = {}
+
+        header = lines[0] # the first line is the header
+        idx = 1
+        while lines[idx] == '-1':
+        	# column names start with fb:row.row. 
+        	curr_line = lines[idx]
+        	column_name_sempre = curr_line[2]
+        	column_idx = curr_line[1]
+        	column_name = column_name_sempre[11:]
+        	column_idx_to_name[column_idx] = column_name
+            idx += 1
+
+        
+        column_node_type_info = [{'string' : 0, 'number' : 0, 'date' : 0} 
+                                 for col in column_idx_to_name]
+        cell_values = set()
+        while idx < len(lines):
+        	curr_line = lines[idx]
+            column_idx = curr_line[1]
+            node_info = zip(header, curr_line)
+            cell_values.add(cls._normalize_string(node_info['content']))  
+            if node_info['date']:
+            	column_node_type_info[column_idx]['date'] += 1 
+            # If cell contains too many tokens, then likely not number
+            elif node_info['number'] and num_tokens < max_tokens_for_num_cell:
+            	column_node_type_info[column_idx]['number'] += 1 
+            else:
+            	column_node_type_info[column_idx]['string'] += 1
+
+
+        return cls(cell_values, column_node_type_info, column_idx_to_name)
+
+
+    def get_entities_from_question(self, question : List[str]):
+        entity_data = []
+        for i, token in enumerate(question):
+            if token in STOP_WORDS: continue
+            normalized_token = self._normalize_string(token)
+            if len(normalized_token) == 0: continue
+            if self._string_in_table(normalized_token):
+                curr_data = {'value' : normalized_token, 'token_start' : i, 'token_end' : i+1}
+                entity_data.append(curr_data)
+                                  
+
+        expanded_entities = self._expand_entities(question, entity_data, table)
+        return expanded_entities #TODO(shikhar) Handle conjunctions
+    
+        
+
+    def _string_in_table(self, candidate : str) -> bool:
+        for cell_value in self.cell_values:
+            if candidate in cell_value: return True
+        return False
+
+
+    def _process_conjunction(self, entity_data):
+        raise NotImplementedError
+
+    def _expand_entities(self, question, entity_data):
```

Comment:
I think I mostly understand the logic here, but writing down a couple of lines explaining it would be very helpful!

Reply:
Definitely, i'll add comments